packages/ai-jsx/src/core/log.ts

```diff
@@ -1,26 +1,61 @@
 import _ from 'lodash';
 import { pino } from 'pino';
 import { Element } from '../index';
-import { v4 as uuidv4 } from 'uuid';
 
 export type LogLevel = 'fatal' | 'error' | 'warn' | 'info' | 'debug' | 'trace';
 
 export type Logger = Record<LogLevel, (obj: object | string, msg?: string) => void>;
 
-/**
- * @param level The logging level.
- * @param element The element from which the log originated.
- * @param renderId A unique identifier associated with the rendering request.
- * @param metadataOrMessage An object to be included in the log, or a message to log.
- * @param message The message to log, if `metadataOrMessage` is an object.
- */
-export type LogImplementation = (
-  level: LogLevel,
-  element: Element<object>,
-  renderId: string,
-  metadataOrMessage: object | string,
-  message?: string
-) => void;
+export abstract class LogImplementation {
+  protected readonly loggedExceptions = new WeakMap<object, boolean>();
+
+  /**
+   * @param level The logging level.
+   * @param element The element from which the log originated.
+   * @param renderId A unique identifier associated with the rendering request.
+   * @param metadataOrMessage An object to be included in the log, or a message to log.
+   * @param message The message to log, if `metadataOrMessage` is an object.
+   */
+  abstract log(
+    level: LogLevel,
+    element: Element<object>,
+    renderId: string,
+    metadataOrMessage: object | string,
+    message?: string
+  ): void;
+
+  /**
+   * Logs exceptions thrown during an element's render. By default invokes `log` with level `"error"`
+   * for the element that threw the exception and level `"trace"` for elements through which the exception
+   * propagated.
+   *
+   * @param element The element from which the exception originated or through which the exception was propagated.
+   * @param renderId A unique identifier associated with the rendering request for this element.
+   * @param exception The thrown exception.
+   */
+  logException(element: Element<object>, renderId: string, exception: any) {
+    let alreadyLoggedException = false;
+    if (typeof exception === 'object' && exception !== null) {
+      alreadyLoggedException = this.loggedExceptions.has(exception);
+      if (!alreadyLoggedException) {
+        this.loggedExceptions.set(exception, true);
+      }
+    }
+
+    const elementTag = `<${element.tag.name}>`;
+    this.log(
+      alreadyLoggedException ? 'trace' : 'error',
+      element,
+      renderId,
+      { exception },
+      `Rendering element ${elementTag} failed with exception: ${exception}`
+    );
+  }
+}
+
+export class EmptyLogImplementation extends LogImplementation {
```

Comment:
```suggestion
export class SilentLogImplementation extends LogImplementation {
```

```suggestion
export class NoOpLogImplementation extends LogImplementation {
```

minor suggestion / do whatever seems best to you